plugins/modules/idrac_diagnostics.py

```diff
@@ -0,0 +1,633 @@
+            export_diagnostics_url = self.__get_export_diagnostics_url(module)
+            job_status = {}
+            self.__check_file_extension(module)
+            if share_type == "local":
+                export_diagnostics_status = self.__export_diagnostics_local(module, export_diagnostics_url)
+            elif share_type in ["http", "https"]:
+                export_diagnostics_status = self.__export_diagnostics_http(module, export_diagnostics_url)
+                job_status = self.get_job_status(module, export_diagnostics_status)
+            elif share_type == "cifs":
+                export_diagnostics_status = self.__export_diagnostics_cifs(module, export_diagnostics_url)
+                job_status = self.get_job_status(module, export_diagnostics_status)
+            elif share_type == "nfs":
+                export_diagnostics_status = self.__export_diagnostics_nfs(module, export_diagnostics_url)
+                job_status = self.get_job_status(module, export_diagnostics_status)
+            status = export_diagnostics_status.status_code
+            if status in self.STATUS_SUCCESS:
+                msg = SUCCESS_EXPORT_MSG
+                job_details = job_status
+                return msg, job_details
+        except HTTPError as err:
+            err_msg = json.loads(err.read())["error"]["@Message.ExtendedInfo"][0]["Message"]
+            if err_msg == EXPORT_NO_RESULTS_MSG:
+                module.exit_json(msg=EXPORT_NO_RESULTS_MSG, skipped=True)
+            module.exit_json(msg=err_msg, failed=True)
+        return export_diagnostics_status
+
+    def __export_diagnostics_local(self, module, export_diagnostics_url):
+        payload = {}
+        payload["ShareType"] = "Local"
+        path = module.params.get('share_parameters').get('share_name')
+        if not (os.path.exists(path) or os.path.isdir(path)):
+            module.fail_json(msg=INVALID_DIRECTORY_MSG.format(path=path), failed=True)
+        if not os.access(path, os.W_OK):
+            module.fail_json(msg=INSUFFICIENT_DIRECTORY_PERMISSION_MSG.format(path=path), failed=True)
+        diagnostics_status = self.__export_diagnostics(module, payload, export_diagnostics_url)
+        diagnostics_file_name = payload.get("FileName")
+        diagnostics_data = self.idrac.invoke_request(diagnostics_status.headers.get("Location"), "GET")
+        diagnostics_output = [line + "\n" for line in diagnostics_data.body.decode().split("\r\n")]
+        file_name = os.path.join(path, diagnostics_file_name)
+        with open(file_name, "w") as fp:
+            fp.writelines(diagnostics_output)
+        return diagnostics_status
+
+    def __export_diagnostics_http(self, module, export_diagnostics_url):
+        payload = {}
+        proxy_details = self.get_proxy_details(module)
+        payload.update(proxy_details)
+        export_status = self.__export_diagnostics(module, payload, export_diagnostics_url)
```

Comment:
Can't we just form the payloads of different shares in methods and the call __export_diagnostics() outside to reduce duplicity.

Reply:
done